pageserver/src/restore_datadir.rs

```diff
@@ -0,0 +1,270 @@
+// is started, it starts streaming from that LSN.
+//
+
+use bytes::{Buf, BytesMut};
+use log::*;
+use regex::Regex;
+use std::env;
+use std::fmt;
+
+use tokio::runtime;
+
+use futures::future;
+
+use crate::{page_cache, PageServerConf};
+use std::fs;
+use walkdir::WalkDir;
+
+pub fn restore_main(conf: &PageServerConf) {
+    // Create a new thread pool
+    let runtime = runtime::Runtime::new().unwrap();
+
+    runtime.block_on(async {
+        let result = restore_chunk(conf).await;
+
+        match result {
+            Ok(_) => {
+                return;
+            }
+            Err(err) => {
+                error!("error: {}", err);
+                return;
+            }
+        }
+    });
+}
+
+async fn restore_chunk(conf: &PageServerConf) -> Result<(), FilePathError> {
+    let pgdata_base_path = env::var("PGDATA_BASE_PATH").unwrap();
+    info!("Restoring from local dir...");
+
+    let sys_id: u64 = 42;
+    let control_lsn = 0; //TODO get it from sysid
+    let mut slurp_futures: Vec<_> = Vec::new();
+
+    for e in WalkDir::new(pgdata_base_path.clone()) {
+        let entry = e.unwrap();
+
+        if !entry.path().is_dir() {
+            let path = entry.path().to_str().unwrap();
+
+            let relpath = path
+                .strip_prefix(&format!("{}/", pgdata_base_path))
+                .unwrap();
+            info!(
+                "Restoring file {} relpath {}",
+                entry.path().display(),
+                relpath
+            );
+
+            let parsed = parse_rel_file_path(&relpath);
+
+            match parsed {
+                Ok(mut p) => {
+                    p.lsn = control_lsn;
+
+                    let f = slurp_base_file(conf, sys_id, path.to_string(), p);
+
+                    slurp_futures.push(f);
+                }
+                Err(e) => {
+                    warn!("unrecognized file: {} ({})", relpath, e);
+                }
+            };
+        }
+    }
+
+    let pcache = page_cache::get_pagecache(conf.clone(), sys_id);
+    pcache.init_valid_lsn(control_lsn);
+
+    info!("{} files to restore...", slurp_futures.len());
+
+    future::join_all(slurp_futures).await;
+    info!("restored!");
+    Ok(())
+}
+
+// From pg_tablespace_d.h
+//
+// FIXME: we'll probably need these elsewhere too, move to some common location
+const DEFAULTTABLESPACE_OID: u32 = 1663;
+const GLOBALTABLESPACE_OID: u32 = 1664;
+
+#[derive(Debug)]
+struct FilePathError {
+    msg: String,
+}
+
+impl FilePathError {
+    fn new(msg: &str) -> FilePathError {
+        FilePathError {
+            msg: msg.to_string(),
+        }
+    }
+}
+
+impl From<core::num::ParseIntError> for FilePathError {
+    fn from(e: core::num::ParseIntError) -> Self {
+        return FilePathError {
+            msg: format!("invalid filename: {}", e),
+        };
+    }
+}
+
+impl fmt::Display for FilePathError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid filename")
+    }
+}
```

Comment:
This can be automatically derived using `thiserror`, for example:

```rust
use thiserror::Error;

#[derive(Debug, Error)]
#[error("invalid filename: {0}")]
struct FilePathError(String);
```

This would also derive the `std::error::Error` trait, which is useful for compatibility with other code.